arrow-libs/core/arrow-core-data/src/main/kotlin/arrow/core/Ior.kt

```diff
@@ -986,45 +912,63 @@ inline fun <A, B, C, D, E, F, G, H, I, J, K, L> Ior<A, B>.zip(
   j: Ior<A, J>,
   k: Ior<A, K>,
   map: (B, C, D, E, F, G, H, I, J, K) -> L
-): Ior<A, L> =
-  if (this is Ior.Right && c is Ior.Right && d is Ior.Right && e is Ior.Right && f is Ior.Right && g is Ior.Right && h is Ior.Right && i is Ior.Right && j is Ior.Right && k is Ior.Right) {
-    Ior.Right(map(value, c.value, d.value, e.value, f.value, g.value, h.value, i.value, j.value, k.value))
-  } else if (this is Ior.Both && c is Ior.Both && d is Ior.Both && e is Ior.Both && f is Ior.Both && g is Ior.Both && h is Ior.Both && i is Ior.Both && j is Ior.Both && k is Ior.Both) {
-    SA.run {
-      Ior.Both(
-        leftValue.combine(c.leftValue).combine(d.leftValue).combine(e.leftValue).combine(f.leftValue)
-          .combine(g.leftValue).combine(h.leftValue).combine(i.leftValue).combine(j.leftValue).combine(k.leftValue),
-        map(
-          rightValue,
-          c.rightValue,
-          d.rightValue,
-          e.rightValue,
-          f.rightValue,
-          g.rightValue,
-          h.rightValue,
-          i.rightValue,
-          j.rightValue,
-          k.rightValue
-        )
-      )
-    }
-  } else SA.run {
+): Ior<A, L> {
+  val rightValue: L? = Nullable.zip(
+    (this@zip as? Right)?.value ?: (this@zip as? Both)?.rightValue,
+    (c as? Right)?.value ?: (c as? Both)?.rightValue,
+    (d as? Right)?.value ?: (d as? Both)?.rightValue,
+    (e as? Right)?.value ?: (e as? Both)?.rightValue,
+    (f as? Right)?.value ?: (f as? Both)?.rightValue,
+    (g as? Right)?.value ?: (g as? Both)?.rightValue,
+    (h as? Right)?.value ?: (h as? Both)?.rightValue,
+    (i as? Right)?.value ?: (i as? Both)?.rightValue,
+    (j as? Right)?.value ?: (j as? Both)?.rightValue,
+    (k as? Right)?.value ?: (k as? Both)?.rightValue,
+    map
+  )
+
+  val leftValue: A? = SA.run {
     var accumulatedLeft: A? = null
-    accumulatedLeft = if (this@zip is Ior.Left) value.maybeCombine(accumulatedLeft) else accumulatedLeft
-    accumulatedLeft = if (c is Ior.Left) c.value.maybeCombine(accumulatedLeft) else accumulatedLeft
-    accumulatedLeft = if (d is Ior.Left) d.value.maybeCombine(accumulatedLeft) else accumulatedLeft
-    accumulatedLeft = if (e is Ior.Left) e.value.maybeCombine(accumulatedLeft) else accumulatedLeft
-    accumulatedLeft = if (f is Ior.Left) f.value.maybeCombine(accumulatedLeft) else accumulatedLeft
-    accumulatedLeft = if (g is Ior.Left) g.value.maybeCombine(accumulatedLeft) else accumulatedLeft
-    accumulatedLeft = if (h is Ior.Left) h.value.maybeCombine(accumulatedLeft) else accumulatedLeft
-    accumulatedLeft = if (i is Ior.Left) i.value.maybeCombine(accumulatedLeft) else accumulatedLeft
-    accumulatedLeft = if (j is Ior.Left) j.value.maybeCombine(accumulatedLeft) else accumulatedLeft
-    accumulatedLeft = if (j is Ior.Left) j.value.maybeCombine(accumulatedLeft) else accumulatedLeft
-    Ior.Left(accumulatedLeft!!)
+    if (this@zip is Left) value.maybeCombine(accumulatedLeft) else accumulatedLeft
```

Comment:
Nit can we define a computational block for this short-circuiting blocks like either.invoke just for Ior

Reply:
Don't think we can since you need to accumulate on `Both` instead of short-circuiting. However, this is a custom implementation for `Ior` since using `flatMap` here results in too long bytecode.